TwoWayDatabinding/app/src/main/java/com/enpassio/twowaydatabinding/ToyRepository.java

```diff
@@ -0,0 +1,53 @@
+package com.enpassio.twowaydatabinding;
+
+import android.arch.lifecycle.LiveData;
+
+import com.enpassio.twowaydatabinding.data.ToyDatabase;
+import com.enpassio.twowaydatabinding.data.ToyEntry;
+import com.enpassio.twowaydatabinding.utils.AppExecutors;
+
+import java.util.List;
+
+public class ToyRepository {
+
+    private static ToyRepository sInstance;
+    private final ToyDatabase mDatabase;
+    private final LiveData<List<ToyEntry>> toyList;
+    private final AppExecutors mExecutors;
+
+    private ToyRepository(ToyDatabase database, AppExecutors executors) {
+        mDatabase = database;
+        mExecutors = executors;
+        toyList = loadToys();
+    }
+
+    public static ToyRepository getInstance(ToyDatabase database, AppExecutors executors) {
+        if (sInstance == null) {
+            synchronized (ToyRepository.class) {
+                sInstance = new ToyRepository(database, executors);
+            }
+        }
+        return sInstance;
+    }
+
+    private LiveData<List<ToyEntry>> loadToys(){
+        return mDatabase.toyDao().getAllToys();
+    }
+
+    public LiveData<List<ToyEntry>> getToyList() {
```

Comment:
I think we can do it with even some less code, still getting the same result. As we're getting the LiveDat<List<Toys>> using `mDatabase.toyDao().getAllToys();` inside the method `loadToys()`, so we can do the same in the current method also. That way we won't require to do extra calculation/operation just when repository is created. But rather we'll do it only when that data is needed the first time. What do you think here @OyaCanli ? Does that strategy sound good to you?

Reply:
Do you mean loading from database inside getter.. Like:
public LiveData<List<ToyEntry>> getToyList(){
if(mToyList == null){
mToyList = mDatabase.toyDao().getAllToys()
}
return mToyList;
Excuse my formatting, I'm on mobile. 
We can do that of course. It seems good to me. Although, since in this sample, the toy list is needed in the host fragment and accessed right away, it won't make much difference.

Reply:
So what I meant is simply like:

```
public class ToyRepository {

    private static ToyRepository sInstance;
    private final ToyDatabase mDatabase;
    private final AppExecutors mExecutors;

    private ToyRepository(ToyDatabase database, AppExecutors executors) {
        mDatabase = database;
        mExecutors = executors;
    }

    public static ToyRepository getInstance(ToyDatabase database, AppExecutors executors) {
        if (sInstance == null) {
            synchronized (ToyRepository.class) {
                sInstance = new ToyRepository(database, executors);
            }
        }
        return sInstance;
    }

    public LiveData<List<ToyEntry>> getToyList() {
        return mDatabase.toyDao().getAllToys();
    }

    public void insertToy(final ToyEntry toy){
        mExecutors.diskIO().execute(() -> mDatabase.toyDao().insertToy(toy));
    }

    public void updateToy(final ToyEntry toy){
        mExecutors.diskIO().execute(() -> mDatabase.toyDao().updateToyInfo(toy));
    }

    public void deleteToy(final ToyEntry toy){
        mExecutors.diskIO().execute(() -> mDatabase.toyDao().deleteToy(toy));
    }

}
```

Reply:
This version looks simpler. Of course the performance is almost the same so we can go with the previous way also but I felt like this was maybe a bit more clear.

Reply:
Hmm. But we wouldn't want to get whole list each time from the database. We can get it once and then observe changes. 
However, we can buffer it inside viewmodel as well. If toyList inside viewmodel is null, claim from repository, which will get from the database..